### PR TITLE
fix(utils): re-implement derive to make glitch free (#335 was buggy)

### PR DESCRIPTION
**Pull request by Noitidart**
Related: https://github.com/pmndrs/valtio/issues/296#issuecomment-1020736885

- [x] add failing test @Noitidart 
- [x] re-implement @dai-shi 
- [x] refactor @dai-shi

### CODE REVIEW

File: tests/derive.test.tsx [hunk 1]
```diff
@@ -303,24 +304,33 @@ it('basic underive', async () => {
...
+        value: computeValue,
+      },
+      { sync: false }
+    )
 
+    let renderCount = 0
     const App = () => {
       const snap = useSnapshot(derived3)
+
+      renderCount++
```
Comment (dai-shi) [inline tests/derive.test.tsx:332]: We should count commits not renders. Let me tweak the test.

File: tests/derive.test.tsx [hunk 2]
```diff
@@ -329,18 +339,17 @@ describe('glitch free', () => {
...
+    await findByText('value: v0: 0, v1: 0, v2: 0')
     expect(computeValue).toBeCalledTimes(1)
+    expect(renderCount).toBe(1)
 
     fireEvent.click(getByText('button'))
-    await findByText('value: 1')
-    expect(computeValue).toBeCalledTimes(2)
+
+    await findByText('value: v0: 1, v1: 1, v2: 1')
+    expect(computeValue).toBeCalledTimes(3)
```
Comment (dai-shi) [inline tests/derive.test.tsx:351]: This should be `2`, right?

**Comment by dai-shi** [comment]
Okay, I kind of new my impl in #335 was truly correct. I had another impl, and based on it, a commit is added. It's complicated, so let me try some refactoring.

Meanwhile, @Noitidart could you try if it solves the issue?

**Comment by dai-shi** [comment]
Let's release this soon, because #342 is a bigger issue. v1.2.10 is full of bugs. 😓

**Review by dai-shi — APPROVED** [review]
Thanks for your investigation. I hope this is better than #335, but not sure if there are some bugs remaining. Please report issues if anything goes wrong.

File: tests/derive.test.tsx (hunk 1 above)
Comment (Noitidart) [reply tests/derive.test.tsx]: Oh very interesting I didn't know useEffect was the way to count commits! I learned a lot here thanks!

**Comment by Noitidart** [comment]
> Okay, I kind of new my impl in #335 was truly correct. I had another impl, and based on it, a commit is added. It's complicated, so let me try some refactoring.
> 
> Meanwhile, @Noitidart could you try if it solves the issue?

Trying now! Sorry for delay I'm on west coast us so just woke up haha.

**Comment by Noitidart** [comment]
With v1.2.9 this used to happen:

```
​compute1 
compute2 
compute3: 
000 
render value: 
v0: 0, v1: 0, v2: 0 
INCREMENT
compute1 
compute3: 
110 
compute2 
compute3: 
111 
render value: 
v0: 1, v1: 1, v2: 1 
```

Notice how compute1 then compute3 happens, then again compute2 happens then compute3 happens.

Now in v1.2.11 its perfect:

```
compute1 
compute2 
compute3: 
000 
render value: 
v0: 0, v1: 0, v2: 0 
INCREMENT
compute1 
compute2 
compute3: 
111 
render value: 
v0: 1, v1: 1, v2: 1 
```

compute1 then compuet2 then compute3 and all variables are lined up! no extra callback is triggered!

**Comment by dai-shi** [comment]
> Trying now! Sorry for delay I'm on west coast us so just woke up haha.

No worries! I knew the timezone difference. I wanted to replace troublesome v1.2.10 as soon as possible.

**Comment by Noitidart** [comment]
Totally gotcha, thanks for giving me chance to take part of the emergency release. If possible it happens again no matter time if I see it I would definitely get on it, thanks for trying to include us man!

**Comment by dai-shi** [comment]
By the way, this implementation might still have a limitation. A circular graph may result in an infinite loop. Would you like to test it??
(It may happen with v1.2.9 too.)

**Comment by Noitidart** [comment]
> By the way, this implementation might still have a limitation. A circular graph may result in an infinite loop. Would you like to test it?? (It may happen with v1.2.9 too.)

Haha thanks! I don't understand the circular graph situation, if you could help me understand I would learn about this circular graph thing.

**Comment by dai-shi** [comment]
The circular graph is something like this: A derived from B derived from C derived from A.

**Comment by dai-shi** [comment]
Found this: https://en.wikipedia.org/wiki/Reactive_programming#Cyclic_dependencies

It's called cyclic dependencies.

**Comment by Noitidart** [comment]
Oh thanks! Sorry I got busy with urgently implementing a Stripe payment system update. I'll have to revisit this please. Thanks very much for trying to get me involved!

**Pull request merged**